operator-common/src/main/java/io/strimzi/operator/common/model/cruisecontrol/CruiseControlHeaders.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.common.model.cruisecontrol;
+
+/**
+ * Enum with Cruise Control headers
+ */
+public enum CruiseControlHeaders {
```

Comment:
Genuine question, what's the need of a specific class for the headers? Why isn't it ok having them in the `CruiseControlApi` interface?

Reply:
This will be shared with the topic operator code, and it's part of the CC public API.

Reply:
So it sounds to me you are creating this enum just for having a central/common place where putting the first user task header because you can't/don't want to move the `CruiseControlApi` interface in the operator common now? This is why I am reasoning if a dedicated headers class is really needed.

Reply:
My plan is to not move the CruiseControlApi, but instead create a new thin CC client based on Java HTTP client, and use that inside CruiseControlApi. This is because CruiseControlApi contains a good amount of the rebalance logic which should not be part of the shared CC client. When all client endpoints will be migrated, we can also think about renaming CruiseControlApi to CruiseControlRebalanceHandler, but that's optional.

Reply:
What do you mean by thin? The current logic in the CruiseControlApi is for handling the response from the CC REST API and translating them for the upper level (see rebalance operator). Is your client just about sending and receiving HTTP requests/responses? Hasn't the topic operator CC interaction the need for handling them?
I am just wondering if we really need two layers.

Reply:
By thin I mean that it will only handle request/response and mapping to POJOs (1:1 mapping with the JSON response), nothing more. This is similar to what Kube client or Kafka admin client do. Instead, CruiseControlApi has some extra logic to map between CC response and what rebalance expects, which makes it hard to maintain and share.

Reply:
> Hasn't the topic operator CC interaction the need for handling them?

Yes, that's why I created a ReplicasChangeHandler that decouples the controller from the CC client. CruiseControlApi will be a similar handler. I think this is a better design overall.

Reply:
I already tried to migrate user_tasks (but I need this PR to move on), and I see that this approach works quite well. As a result, the CruiseControlApi logic is much easier to read.

Reply:
Yeah, mapping stuff, I mean that. 

I think it's better to keep that logic isolated in a specific handler (CruiseControlApi in this case). Every component that requires CC integration also needs some kind of custom mapping logic. For example, the topic operator doesn't need the the full response when calling user_tasks, which makes for a faster call.

Additionally CruiseControlResponse depends on Vertx, which would also require RebalanceAssemblyOperator refactoring. This can also be done to get rid of Vertx, but I think should be a follow up PR to not over complicate this task.

Reply:
I am not totally sure I agree with that. You would like to have one HTTP client which sends requests and get responses from Cruise Control API but then passing to an upper layer the responsability to decode and building a POJO for the response. This layer could be different between cluster operator (because of rebalance) and topic operator (because of replication factor change). You could easily end up with duplicated code because both could need most of the same fields from the received JSON. Why not having one and making a POJO which could be a common denominator for the needs, not adding ALL fields from JSON response but the ones who are useful for our needs on both cluster and topic operator. In the end across the two it's the the user task request to be in common, or? Not sure what the other @strimzi/maintainers think about this.

Reply:
> Why not having one and making a POJO which could be a common denominator for the needs, not adding ALL fields from JSON response but the ones who are useful for our needs on both cluster and topic operator.

In my view, the shared Cruise Control client shouldn't be concerned with response handling. This client should simply send the request and return a response POJO. Much like Kafka admin and Kube clients do. 

The response handling is specific to the functionality we are implementing (the same response could be processed in different ways), so it should be the responsibility of the caller (CruiseControlApi in CO, ReplicasChangeHandler in TO).

That way, each object involved in the call chain trigger by the controller is focused on doing one thing and the code is easier to read and maintain.

This would leave a Vertx dependency in CruiseControlApi required by KafkaRebalanceAssemblyOperator. Of course we can also get rid of that, but I would leave it for a dedicated PR.

Reply:
> In my view, the shared Cruise Control client shouldn't be concerned with response handling. This client should simply send the request and return a response POJO. Much like Kafka admin and Kube clients do.

Here it depends on what you mean by "response handling". In my mind the Cruise Control client has to do what you are saying, so using an HTTP client to send a request, get a response and from that building a POJO. At the same time this is what the current `CruiseControlApiImpl` is doing AFAICS. It sends an HTTP request, get the response and put the received JSON into a generic `CruiseControlResponse` POJO (or into a more specific `CruiseControlRebalanceResponse` POJO, depending on the call). So it's mapping the HTTP response body into a POJO as you say. Then the upper level, the `KafkaRebalanceAssemblyOperator` is taking care of extracting from the POJO what it needs. I don't see in the client any rebalancing logic.

> The response handling is specific to the functionality we are implementing (the same response could be processed in different ways), so it should be the responsibility of the caller (CruiseControlApi in CO, ReplicasChangeHandler in TO).

This is where I don't agree. What does it mean `the same response could be processed in different ways` ... building/filling a different POJO for the same HTTP response? I don't agree. Building/filling part of a POJO depending on which part is calling the client? I still don't agree.

As described above, which seems to be what we have today (that could be improved of course), the Cruise Control client should build a POJO from the HTTP response and this POJO is handled by topic operator or cluster operator (rebalance operator). Each of them will get what they need from the POJO. But it will be always the same POJO for both because it's mapped on a corresponding HTTP response.

Example: the user task endpoint returns an HTTP response. It should be mapped to a unique UserTask POJO usable by both topic and cluster operator and not having two different handlers as you say to fill different POJOs.

Reply:
> In my mind the Cruise Control client has to do what you are saying, so using an HTTP client to send a request, get a response and from that building a POJO. At the same time this is what the current CruiseControlApiImpl is doing AFAICS.

Not exactly. It's not a one to one mapping between CC response fields and POJO properties, but a custom mapping in a format that is suitable for the rebalance operator. 

As an example, the user_tasks response has many fields, but we only have two of them in the CruiseControlResponse (userTaskId and originalResponse). Additionally, the orginalResponse is parsed into a JsonObject, which causes extra delay that may not be needed by other uses cases.

> What does it mean the same response could be processed in different ways ... building/filling a different POJO for the same HTTP response? I don't agree. Building/filling part of a POJO depending on which part is calling the client? I still don't agree.

Absolutely not. Each response has it's own POJO, which contains all the fields we get from the server. Then the caller decides which fields to use, depending on the use case. Of course, we have some optional request flags that determines the presence or not of some extra content. This will be reflected in the client API.

> As described above, which seems to be what we have today (that could be improved of course), the Cruise Control client should build a POJO from the HTTP response and this POJO is handled by topic operator or cluster operator (rebalance operator). Each of them will get what they need from the POJO. But it will be always the same POJO for both because it's mapped on a corresponding HTTP response. 

> Example: the user task endpoint returns an HTTP response. It should be mapped to a unique UserTask POJO usable by both topic and cluster operator and not having two different handlers as you say to fill different POJOs.

That's basically what I'm proposing here. The current client is centered around Vertx, so it requires a rewrite.

Reply:
> Not exactly. It's not a one to one mapping between CC response fields and POJO properties, but a custom mapping in a format that is suitable for the rebalance operator.

We didn't add all the fields because we didn't need them. It's just a matter of extending with more fields the topic operator would need. Doing a full mapping with fields that won't be used doesn't really make sense imho. We could add other fields in the POJO in the future.

> In my proposal, each response has it's own POJO, which contains all the fields we get from the server. Then the caller decides which fields to use, depending on the use case. Of course, we have some optional request flags that determines the presence or not of some extra content. This will be reflected in the client API.

As said we can avoid having everything but the minimum needed by the different components who use the Cruise Control API. Anyway, it's different from what you said or I understood before: having a different mapping to a different POJO from the same HTTP response, depending on who is going to use that POJO. That was what I didn't like. 

So AFAICS, the `CruiseControlApi` is already doing what we need. It could be improved (removing Vert.x as you said) but it's already mapping HTTP response to POJO where POJO doesn't mean exactly having all fields but bringing a JSON and the component using the API gets the fields it needs from the JSON itself. You could even have a strong typed POJO if you want instead of bringing POJO. 

Anyway I think that this change/refactoring needs a proposal.

Reply:
You mean the long term goal of sharing the CC client right? I will raise a proposal for that. Instead, what do you think about this PR?

I explained the motivation around creating the Headers enum [here](https://github.com/strimzi/strimzi-kafka-operator/pull/10092#discussion_r1601290240), but I can revert that change if you think it's not worth.

Reply:
> Anyway I think that this change/refactoring needs a proposal.

Agree. Thanks.